src/Voron/Data/Tables/TableValueCompressor.cs

```diff
@@ -229,6 +229,18 @@ private void MaybeTrainCompressionDictionary(Table table, FixedSizeTree etagsTre
                         }
 
                         tx.LowLevelTransaction.OnDispose += RecreateRecoveryDictionaries;
+                        tx.LowLevelTransaction.OnDispose += state =>
+                        {
+                            if (state is not LowLevelTransaction llt || llt.Committed)
+                                return;
+                            // ***************************************
+                            // RavenDB-17758  - This is *important* - if we aren't committing the transaction, we *have*
+                            // to remove the in memory cache for the dictionary, since it wasn't written to disk. If we
+                            // keep it, we can compress data with a dictionary that we rolled back, and end up with data
+                            // corruption!!!
+                            // ****************************************
+                            tx.LowLevelTransaction.Environment.CompressionDictionariesHolder.Remove(newId);
```

Comment:
`OnDispose` is called _after_ we release the write transaction lock here:

https://github.com/ravendb/ravendb/blob/559653d13f3dce7f05098d34bac67b4d8a8110ce/src/Voron/StorageEnvironment.cs#L912

So at the time of calling `Remove()` here we might have a new write transaction running. Don't we need to remove it from the cache _before_ creating a new write tx?

Reply:
Is it possible to have a new (write) transaction come in at that time frame?
Read transactions cannot observe the new id, since there is no data for it (was rolled back).

I agree that this is a concern, however. We can add that to the rollback, I guess, or better yet, add an event there?

Reply:
I like the solution with adding `OnRollBack` event and handling it there (as implemented in https://github.com/ravendb/ravendb/pull/13373/commits/15040a028f7d285104362852d2ea179ff6ac3871)